app/models/submission.rb

```diff
@@ -255,6 +255,52 @@ def make_remark_result
     remark.save
   end
 
+  def copy_grading_data(old_submission)
+    return if old_submission.blank?
+
+    # copy submission-wide data to this submission
+    old_submission.feedback_files.each do |feedback_file|
+      feedback_file_dup = feedback_file.dup
+      feedback_file_dup.update!(submission_id: self.id)
+    end
+
+    old_submission.test_runs.each do |test_run|
+      test_run_dup = test_run.dup
+      test_run_dup.update!(submission_id: self.id)
+
+      test_run.test_group_results.each do |test_group_result|
+        test_group_result_dup = test_group_result.dup
+        test_group_result_dup.update!(test_run_id: test_run_dup.id)
+
+        test_group_result.test_results.each do |test_result|
+          test_result_dup = test_result.dup
+          test_result_dup.update!(test_group_result_id: test_group_result_dup.id)
+        end
+
+        test_group_result.feedback_files.each do |feedback_file|
+          feedback_file_dup = feedback_file.dup
+          feedback_file_dup.update!(test_group_result_id: test_group_result_dup.id)
+        end
+      end
+    end
+
+    self.get_original_result.copy_grading_data(old_submission.get_original_result)
+
+    # whether it's been submitted or not, remark request is going to be copied
+    if old_submission.remark_request.present?
+      self.remark_request = old_submission.remark_request
+      self.remark_request_timestamp = old_submission.remark_request_timestamp
+      self.save!
+
+      self.create_result # additional result for remark data
+
+      # if there's already a remark result created as well
+      if old_submission.has_remark?
```

Comment:
This check is redundant with the `old_submission.remark_request.present?` check above (unless I'm missing something?)

Reply:
So this is a bit odd, even to me - I'm overall trying to handle retaining data from two cases here:

1. In the case where the student is writing the remark request, has saved it, but has not submitted it to the instructor yet, the `remark_request` attribute and associated timestamp is populated on the submission, but no remark _result_ has been created yet; so I just check the attribute manually.
2. Where the student has _submitted_ the remark request to the instructor, everything in case 1 happens, but also a new remark result _is_ created, so we have to run the copy on that. Since `has_remark?` basically checks for if that remark result exists or not, I used it here to verify this case.

One thing worth noting is that `has_remark?` has the following comments:

> Returns a remark result that may or may not have been submitted. If you want the submitted remark result then the submitted_remark association should be used instead.

However, unless I'm misinterpreting what it means for a remark to not be submitted yet, I don't think this is fully accurate. From what I observed, I'm not actually sure it's possible to have a remark result without submitting it. To me, the function `has_remark?` seems to do practically the same thing as `submitted_remark`, but in a different way. Specifically, they both return the remark result (which I'm assuming is always the latest result if there is more than one result on the submission); `submitted_remark` gets it using the `remark_request_submitted_at` timestamp on `Result` and `has_remark` returns the last result directly _contingent_ on there being 2 results minimum). Wasn't sure which one to use so I just went with `has_remark?`, but `submitted_remark` might be better for readability.

Do you have any guidance on the best way to handle both of the above cases (or if I am even interpreting this correctly)?

Reply:
Ah I see, yes that makes sense. In this case I believe you can do the following:

1. *Always* copy the `remark_request` and `remark_request_timestamp` attributes from the old submission to the new submission; this doesn't need to be guarded by a conditional.
2. Move the `self.create_result` to inside the `old_submission.has_remark?` check. You can also simplify a bit by first doing something like `old_remark_result = old_submission.remark_result` and then checking whether `old_remark_result.present?`